### PR TITLE
Generator tests

### PR DESCRIPTION
**Pull request by jastice**
Includes previous PRs #82  #83 

I created ScalaCheck generators for all the bsp4j data classes and roundtrip serialization tests bsp4j -> bsp4s -> bsp4j

While writing the tests, I found a few inconsistencies between the APIs and spec. These are fixed in this PR.

### CODE REVIEW

File: docs/bsp.md [hunk 1]
```diff
@@ -717,6 +736,10 @@ trait SourceItem {
     * directory belongs to this source item.
     */
   def uri: Uri
+  
+  /** Specifies whether uri denotes a directory or regular file. 
+    * If this is not set, and the uri does not point to an existing file, it should be treated as directory by default. */
+  def isDirectory: Option[Boolean]
```
Comment (olafurpg) [inline docs/bsp.md]: Are we sure we want to use `isDirectory` instead of `kind`?

Comment (jastice) [reply docs/bsp.md]: not feeling strongly, but as stated in #83 I think this is a case where we needn't expect expansion of types of files on this dimension :)

**Comment by jvican** [comment]
I'll review this as soon as the other PRs are acted on and merged, to simplify the review of this PR :smile:

**Comment by jastice** [comment]
I rebased onto the other PRs. Please have a look, but don't merge yet, I'm still updating as I write tests on IntelliJ side.

**Comment by jastice** [comment]
Okay I consider it ready for merge after review. Please have another look @olafurpg @jvican

**Comment by jastice** [comment]
> Note that this PR has several binary breaking changes. It's not a problem since we're in milestone phase but once we commit to a stable v2 we should avoid breaking changes.

absolutely. Part of the point of this was to suss out problems before setting on a stable v2. Which ideally would happen next week to make the IntelliJ bsp support public.

**Comment by olafurpg** [comment]
I agree, it's fantastic to iron out so many issues before the stable release 👍

**Pull request merged**